### PR TITLE
feat: migrate to ggml main

### PR DESCRIPTION
**Pull request by tyrauber**
Correct me if I am wrong, but the ggml fork is not required.

This PR migrates to the official ggml submodule, and extracts the handful of tts methods required to src/ggml-tts-ext.cpp.

Tested on the cli with both variants of the Dia Q4 models you provided.

Now the library can keep current with ggml and benefit from recent changes. 

Should help with performance and platform support.

Thank you very much for this library.

### CODE REVIEW

File: src/util.h [hunk 1]
```diff
@@ -1,63 +1,84 @@
...
+                               bool include_split_characters = false);
+std::vector<std::string> split(std::string target, const char split_on,
+                               bool include_split_characters = false);
 std::string strip(std::string target, std::string vals = " ");
-std::string replace_any(std::string target, std::string to_replace, std::string replacement);
+std::string replace_any(std::string target, std::string to_replace,
+                        std::string replacement);
 
-[[noreturn]] void tts_abort(const char * file, int line, const char * fmt, ...);
+[[noreturn]] void tts_abort(const char *file, int line, const char *fmt, ...);
```
Comment (danielzgtg) [inline src/util.h:82]: Why was this spacing changed away from the ggml-org code style?

Comment (tyrauber) [reply src/util.h:82]: @danielzgtg, that's a good question. Wasn't intentional.  I'll see if I can revert that.

**Comment by tyrauber** [comment]
Actually, looks like this missed metal support. Marked this PR as a draft until resolved.

**Comment by danielzgtg** [comment]
Thank you for your concern in allowing us to "keep current with ggml and benefit from recent changes."

We were originally planning to [rebase on top of latest ggml](https://github.com/mmwillet/ggml/pull/1#issuecomment-2847923270). As there wasn't any pressing need to do so, I didn't do that yet. In the long run, I hope to upstream stuff like `ggml_squeeze` once the code is clean enough.

Soon (tm), I need the following patch in ggml to fix a linker error:

```patch
diff --cc src/ggml-hip/CMakeLists.txt
index fccf8eb,fccf8eb..11e269a
--- a/src/ggml-hip/CMakeLists.txt
+++ b/src/ggml-hip/CMakeLists.txt
@@@ -7,6 -7,6 +7,8 @@@ if (NOT EXISTS $ENV{ROCM_PATH}
  else()
      set(ROCM_PATH $ENV{ROCM_PATH})
  endif()
++add_link_options(-fPIE)
++add_compile_options(-fPIE)
  
  list(APPEND CMAKE_PREFIX_PATH  ${ROCM_PATH})
  list(APPEND CMAKE_PREFIX_PATH "${ROCM_PATH}/lib64/cmake")
```

Is there any way mmwillet/ggml can be updated? I'm somewhat open to merging this PR into TTS.cpp so we can have the small benefit of accessing my recently merged ggml backtrace PR. However, we will need to fork again if there is a need to add or modify HIP/CUDA code.

**Comment by tyrauber** [comment]
I am running to issues with the other models.  Specifically, the Kokoro gguf linked in the readme.

The forked mmwillet/ggml supports conv_transpose_1d operations with non-zero padding values
The main ggml-org/ggml repository has a strict assertion: GGML_ASSERT(p0 == 0) that requires padding=0 for conv_transpose_1d operations. Kokoro models use conv_transpose_1d layers with padding=1 in their generator upsample layers.

I think py-gguf/tts_encoders/kokoro_gguf_encoder.py should be adjusted with Bias Tensor Adjustment,  that automatically resizes bias tensors when conv_transpose_1d padding changes from non-zero to zero, making kokoro gguf models compatible with ggml main.  Thoughts?

**Comment by danielzgtg** [comment]
> other models. Specifically, the Kokoro gguf

At this point that's the main model and basically the only fast-enough model.

> ggml-org/ggml repository has a strict assertion

That assertion means "not yet implemented", or else the function parameter `p0` should be removed and inline because it'd be a constant otherwise. This feature should be implemented in upstream.

> kokoro_gguf_encoder.py should be adjusted with Bias Tensor Adjustment

I'd rather not break compatibility with existing files. We don't have versioning and I can't unpublish other peoples' Hugging Face existing models that are linked in various places.

What won't break compatibility is this: We can do any fixup in the `assign_weights` functions of each model .cpp. This'd reduce the percent available for a future mmap feature but whatever. We're already doing transposes and other stride tricks. I'm also thinking of doing the "[increment the gamma bias by one](https://github.com/mmwillet/TTS.cpp/blob/ec35075bbe79a4512a6bae85067029ffe918056f/src/kokoro_model.cpp#L94)" at load-time like this.

**Comment by mmwillet** [comment]
@tyrauber Sorry for my absence. Currently, the ggml fork is required, as there are several major modifications made in order to properly support various behaviors. I'll work to upgrade the patch this week and update documentation to reflect this.

**Comment by mmwillet** [comment]
> am trying to figure out a workaround in ggml-tts-ext.h to handle this discrepancy without having to fork and patch ggml.

One possible solution to avoid having to patch ggml entirely would be to reimplement added operations as custom map operations. I chose to avoid this initially because the custom map function is limited, has historically been bug prone, and is not compatible with acceleration. Let me see if I can hack together a working work around this week.

**Comment by mmwillet** [comment]
@tyrauber 

>Is there any documentation on the specific requirements that necessitated the ggml fork? I'm trying to understand what functionality is required for TTS that isn't provided in standard ggml.

No, I have not put together any documentation on this outside of scattered tickets and/or inline comments. I will do so on the primary README later today.

In terms of what the modifications are I think you mostly outlined them, but I'll reiterate and amend here for posterity:

> - Conv transpose 1D with non-zero padding - Standard ggml's conv_transpose_1d enforces p0 == 0, but the fork removes this constraint

As noted by yourself, the biggest gap stems from the native GGML convolution transposition implementation which doesn't support full parity with Tensorflow or PyTorch. Besides not supporting non-zero padding or output padding, it also doesn't support grouping which is used by Kokoro (though the support that I added for grouping is still somewhat incomplete).

> - Custom STFT/ISTFT implementations - Audio processing operations for magnitude/phase handling

To a certain degree, it is not necessary to support these operations as part of the graph (as I recall they are not implemented as part of the graph in llama.cpp) and it should be possible to support these operations in their current state via GGML's custom_map ops. Acceleration of these operations would however require that they are independently implemented as their own GGML ops (e.g. good threading support and shader support for metal, cuda, etc).

> - Linear upscaling for 1D sequences - Different from standard nearest-neighbor upscaling

This could currently be a custom map operation and it is unlikely that implementing this as such would result in any significant performance degradation even when accelerating the model via CUDA/metal ops (it is so fast and can just be piped to the cpu with little time loss). My implementation of it is also sloppy. For better parity with PyTorch or Tensorflow, it really should be implemented as a passable option to the generic upscaling operation (along with support for 'bilinear', 'bicubic', 'trilinear', 'area', and 'nearest-exact' modes as well) rather than its own operation.

> - Metal acceleration support for these custom operations

This and CUDA + other shader-like acceleration protocols was the main long term motivator to implement these as native (rather than custom) operations, but I have failed to make the headway that I expected to. 

>I really think it is important to identify exactly what ggml needs to support tts and get that merged upstream.

I agree, this would 100% be the best outcome.

**Comment by mmwillet** [comment]
@tyrauber @danielzgtg [I'm going to pull further discussion into the issue here.](https://github.com/mmwillet/TTS.cpp/issues/66)

**Comment by mmwillet** [comment]
@tyrauber Thanks for your input on this. I am going to close this PR for now, but happy to reopen if our approach changes and we need to return to it.

**Comment by tyrauber** [comment]
@mmwillet, I got a branch where the custom tts ggml methods are extracted to an ext class, but if we are upstreaming - which is definitely the right approach - I don't think you need it. If helpful let me know.   

When you get around to it, some documentation around tts ggml requirements would be really helpful. I still don't understand why tts requires a zero pad conv.  Why did the Dia model work without that? Is that because different tts models have different training, layers, etc.  I am sure there is a reason, I just don't grok it yet.  If you could braindump how ggml tts works, and why it requires unconventional padding and grouping, I think it would be hugely beneficial. Help contributors get up to speed. Right now, it's a lot of trial and error, and printf.

Thank you!  Fantastic library. Really appreciate your and @danielzgtg's responsiveness and willingness to discuss and share. Mad respect.

**Comment by tyrauber** [comment]
llama.rn just added TTS multi-modal support using OuteTTS.  llama.rn is a react-native wrapper around ggml-org/llama.cpp. llama.cpp is using a custom version of ggml. I think active development of ggml is happening there.  The ggml repo mentions "some of the development is currently happening in the llama.cpp and whisper.cpp repos."  

@mmwillet, @danielzgtg, Have either of you checked to see if llama.cpp/ggml has the required TTS methods? 

Perhaps, instead of switching to ggml main, it would make sense to switch to llama.cpp, where active development of ggml is happening?
